### ISSUE TITLE
script frozen with pyenv python : sys.version shows system python version??

### ISSUE THREAD

**Issue by elguavas** [marcelotduarte/cx_Freeze#701]
pyenv installed python 3.8.4 , cx-freeze 6.2

hi there, if i freeze even a basic script using a pyenv installed version of python, when the script runs it reports sys.version of the system python... why??

cxfreeze if definitely being run from the pyenv installed python 3.8.4, on my system the system (non-pyenv) python version is 3.8.2, when i run the frozen script, sys.version reports 3.8.2 (system python) not 3.8.4 (the pyenv installed version of python the script is frozen with).

trivial example script:

import sys

print('sys.version', sys.version)
print('sys.version_info', sys.version_info)

**Comment by marcelotduarte** [comment]
Check this:
https://stackoverflow.com/questions/38772946/what-exactly-does-configure-enable-shared-do-during-python-altinstall
and this:
https://bugs.python.org/issue27685

**Comment by elguavas** [comment]
> Check this:
> https://stackoverflow.com/questions/38772946/what-exactly-does-configure-enable-shared-do-during-python-altinstall
> and this:
> https://bugs.python.org/issue27685

ahh i see. so this means the issue is with my pyenv build is that that it building with --enable-shared and that is pulling in the system python3 libs?

so i should somehow force the pyenv build process to not use --enable-shared?

does cx-freeze require a python built with --enable-shared?

thanks for your time marcel.

**Comment by elguavas** [comment]
hmm, i just checked and it seems pyenv does not build with --enable-shared by default... 

does this mean my pyenv build is not causing the issue then, or does it mean cx-freeze does require shared libs and is picking up the system ones since there are no pyenv build ones? 

... i can see the problem described in the links you posted marcel, but i'm confusing myself over how to solve the problem for my use case. :|

**Comment by marcelotduarte** [comment]
What is your system? Linux, Mac?

**Comment by elguavas** [comment]
> What is your system? Linux, Mac?

sorry, this is on ubuntu 20.04.

can you confirm that cx-freeze needs shared libraries?

**Comment by elguavas** [comment]
to be clear i can easily force pyenv to build with --enable-shared if that is required by cx-freeze. 

once i get this working as it should i will post the full resolution here before closing the issue so any searches might find it. ;)

**Comment by marcelotduarte** [comment]
cx-freeze needs to find the shared library to copy to build folder. In some systems, Python is static linked and distribute the dynamic lib (See this: https://github.com/conda-forge/python-feedstock/issues/222)
I was looking for information due to a problem with macOS and I saw that the information served to explain your case.
I too use Ubuntu 20.04 as a primary machine.

**Comment by elguavas** [comment]
hmm, i just looked in my pyenv build of 3.8.4 and it has a 'lib-dynload' directory full of *.so even though pyenv says it does not build shared libs by default... ehh, my head is spinning now.

maybe i should force pyenv to rebuild 3.8.4 using --enable-shared and then do more testing?

**Comment by marcelotduarte** [comment]
Check with
from distutils import sysconfig
sysconfig.get_config_var('Py_ENABLE_SHARED')

**Comment by marcelotduarte** [comment]
I have a develop version that try to force the copy of theses libraries. Try it out.
pip install -U git+https://github.com/marcelotduarte/cx_Freeze.git@develop

**Comment by elguavas** [comment]
ok, so i rebuilt my pyenv python 3.8.4 with --enable-shared and verified via sysconfig that it was a shared libs python.

first i cx-froze the trivial example script above with regular pip cx-freeze installed in my python 3.8.4. this still resulted in the same problem originally reported, sys.version reported the system python version.

second i pip installed the devel version of cx-freeze you gave above. this resulted in problem solved! sys.version from the frozen executable reported the correct python 3.8.4 that i was running the freeze from.

so, whatever you have changed in that development version makes this problem go away.  if these changes do not impact the normal operation of cx-freeze will they be rolled into a regular pip release soon??

thanks.

**Comment by marcelotduarte** [comment]
Good to hear. Soon I will release a version. I warn you.

**Comment by elguavas** [comment]
hmm, then i immediately ran into a new, different, problem.

using the devel cx-freeze you linked to on my trivial example script worked fine, but as soon as i tried to use it to freeze the larger, much more complex, project i am working on cx-freeze crashed with an AttributeError:

...
File "/home/smg/.pyenv/versions/quill_3.8.4/lib/python3.8/site-packages/importlib_metadata/__init__.py", line 417, in find_distributions
found = self._search_paths(context.pattern, context.path)
AttributeError: 'Context' object has no attribute 'pattern'

i will paste the full traceback below. does this need to be a new issue against the devel version??

full traceback:

running build_exe
Traceback (most recent call last):
  File "__sg_pypr_build_frozen.py", line 100, in <module>
    setup(#name=app.title,
  File "/home/smg/.pyenv/versions/quill_3.8.4/lib/python3.8/site-packages/cx_Freeze/dist.py", line 342, in setup
    distutils.core.setup(**attrs)
  File "/home/smg/.pyenv/versions/3.8.4/lib/python3.8/distutils/core.py", line 148, in setup
    dist.run_commands()
  File "/home/smg/.pyenv/versions/3.8.4/lib/python3.8/distutils/dist.py", line 966, in run_commands
    self.run_command(cmd)
  File "/home/smg/.pyenv/versions/3.8.4/lib/python3.8/distutils/dist.py", line 985, in run_command
    cmd_obj.run()
  File "/home/smg/.pyenv/versions/quill_3.8.4/lib/python3.8/site-packages/cx_Freeze/dist.py", line 217, in run
    freezer.Freeze()
  File "/home/smg/.pyenv/versions/quill_3.8.4/lib/python3.8/site-packages/cx_Freeze/freezer.py", line 646, in Freeze
    self.finder = self._GetModuleFinder()
  File "/home/smg/.pyenv/versions/quill_3.8.4/lib/python3.8/site-packages/cx_Freeze/freezer.py", line 384, in _GetModuleFinder
    finder.IncludePackage(name)
  File "/home/smg/.pyenv/versions/quill_3.8.4/lib/python3.8/site-packages/cx_Freeze/finder.py", line 664, in IncludePackage
    module = self._ImportModule(name, deferredImports)
  File "/home/smg/.pyenv/versions/quill_3.8.4/lib/python3.8/site-packages/cx_Freeze/finder.py", line 311, in _ImportModule
    module = self._InternalImportModule(name,
  File "/home/smg/.pyenv/versions/quill_3.8.4/lib/python3.8/site-packages/cx_Freeze/finder.py", line 412, in _InternalImportModule
    module = self._LoadModule(name, fp, path, info, deferredImports,
  File "/home/smg/.pyenv/versions/quill_3.8.4/lib/python3.8/site-packages/cx_Freeze/finder.py", line 425, in _LoadModule
    return self._LoadPackage(name, path, parent, deferredImports,
  File "/home/smg/.pyenv/versions/quill_3.8.4/lib/python3.8/site-packages/cx_Freeze/finder.py", line 497, in _LoadPackage
    self._LoadModule(name, fp, path, info, deferredImports, parent)
  File "/home/smg/.pyenv/versions/quill_3.8.4/lib/python3.8/site-packages/cx_Freeze/finder.py", line 476, in _LoadModule
    self._ScanCode(module.code, module, deferredImports)
  File "/home/smg/.pyenv/versions/quill_3.8.4/lib/python3.8/site-packages/cx_Freeze/finder.py", line 591, in _ScanCode
    importedModule = self._ImportModule(name, deferredImports,
  File "/home/smg/.pyenv/versions/quill_3.8.4/lib/python3.8/site-packages/cx_Freeze/finder.py", line 346, in _ImportModule
    module = self._InternalImportModule(name,
  File "/home/smg/.pyenv/versions/quill_3.8.4/lib/python3.8/site-packages/cx_Freeze/finder.py", line 412, in _InternalImportModule
    module = self._LoadModule(name, fp, path, info, deferredImports,
  File "/home/smg/.pyenv/versions/quill_3.8.4/lib/python3.8/site-packages/cx_Freeze/finder.py", line 476, in _LoadModule
    self._ScanCode(module.code, module, deferredImports)
  File "/home/smg/.pyenv/versions/quill_3.8.4/lib/python3.8/site-packages/cx_Freeze/finder.py", line 591, in _ScanCode
    importedModule = self._ImportModule(name, deferredImports,
  File "/home/smg/.pyenv/versions/quill_3.8.4/lib/python3.8/site-packages/cx_Freeze/finder.py", line 346, in _ImportModule
    module = self._InternalImportModule(name,
  File "/home/smg/.pyenv/versions/quill_3.8.4/lib/python3.8/site-packages/cx_Freeze/finder.py", line 412, in _InternalImportModule
    module = self._LoadModule(name, fp, path, info, deferredImports,
  File "/home/smg/.pyenv/versions/quill_3.8.4/lib/python3.8/site-packages/cx_Freeze/finder.py", line 425, in _LoadModule
    return self._LoadPackage(name, path, parent, deferredImports,
  File "/home/smg/.pyenv/versions/quill_3.8.4/lib/python3.8/site-packages/cx_Freeze/finder.py", line 497, in _LoadPackage
    self._LoadModule(name, fp, path, info, deferredImports, parent)
  File "/home/smg/.pyenv/versions/quill_3.8.4/lib/python3.8/site-packages/cx_Freeze/finder.py", line 476, in _LoadModule
    self._ScanCode(module.code, module, deferredImports)
  File "/home/smg/.pyenv/versions/quill_3.8.4/lib/python3.8/site-packages/cx_Freeze/finder.py", line 591, in _ScanCode
    importedModule = self._ImportModule(name, deferredImports,
  File "/home/smg/.pyenv/versions/quill_3.8.4/lib/python3.8/site-packages/cx_Freeze/finder.py", line 346, in _ImportModule
    module = self._InternalImportModule(name,
  File "/home/smg/.pyenv/versions/quill_3.8.4/lib/python3.8/site-packages/cx_Freeze/finder.py", line 412, in _InternalImportModule
    module = self._LoadModule(name, fp, path, info, deferredImports,
  File "/home/smg/.pyenv/versions/quill_3.8.4/lib/python3.8/site-packages/cx_Freeze/finder.py", line 476, in _LoadModule
    self._ScanCode(module.code, module, deferredImports)
  File "/home/smg/.pyenv/versions/quill_3.8.4/lib/python3.8/site-packages/cx_Freeze/finder.py", line 591, in _ScanCode
    importedModule = self._ImportModule(name, deferredImports,
  File "/home/smg/.pyenv/versions/quill_3.8.4/lib/python3.8/site-packages/cx_Freeze/finder.py", line 311, in _ImportModule
    module = self._InternalImportModule(name,
  File "/home/smg/.pyenv/versions/quill_3.8.4/lib/python3.8/site-packages/cx_Freeze/finder.py", line 412, in _InternalImportModule
    module = self._LoadModule(name, fp, path, info, deferredImports,
  File "/home/smg/.pyenv/versions/quill_3.8.4/lib/python3.8/site-packages/cx_Freeze/finder.py", line 425, in _LoadModule
    return self._LoadPackage(name, path, parent, deferredImports,
  File "/home/smg/.pyenv/versions/quill_3.8.4/lib/python3.8/site-packages/cx_Freeze/finder.py", line 497, in _LoadPackage
    self._LoadModule(name, fp, path, info, deferredImports, parent)
  File "/home/smg/.pyenv/versions/quill_3.8.4/lib/python3.8/site-packages/cx_Freeze/finder.py", line 476, in _LoadModule
    self._ScanCode(module.code, module, deferredImports)
  File "/home/smg/.pyenv/versions/quill_3.8.4/lib/python3.8/site-packages/cx_Freeze/finder.py", line 591, in _ScanCode
    importedModule = self._ImportModule(name, deferredImports,
  File "/home/smg/.pyenv/versions/quill_3.8.4/lib/python3.8/site-packages/cx_Freeze/finder.py", line 346, in _ImportModule
    module = self._InternalImportModule(name,
  File "/home/smg/.pyenv/versions/quill_3.8.4/lib/python3.8/site-packages/cx_Freeze/finder.py", line 412, in _InternalImportModule
    module = self._LoadModule(name, fp, path, info, deferredImports,
  File "/home/smg/.pyenv/versions/quill_3.8.4/lib/python3.8/site-packages/cx_Freeze/finder.py", line 427, in _LoadModule
    module = self._AddModule(name, file_name=path, parent=parent)
  File "/home/smg/.pyenv/versions/quill_3.8.4/lib/python3.8/site-packages/cx_Freeze/finder.py", line 175, in _AddModule
    module = self._modules[name] = Module(name, path, file_name,
  File "/home/smg/.pyenv/versions/quill_3.8.4/lib/python3.8/site-packages/cx_Freeze/finder.py", line 725, in __init__
    requires = importlib_metadata.requires(packages[0])
  File "/home/smg/.pyenv/versions/3.8.4/lib/python3.8/importlib/metadata.py", line 566, in requires
    return distribution(distribution_name).requires
  File "/home/smg/.pyenv/versions/3.8.4/lib/python3.8/importlib/metadata.py", line 504, in distribution
    return Distribution.from_name(distribution_name)
  File "/home/smg/.pyenv/versions/3.8.4/lib/python3.8/importlib/metadata.py", line 172, in from_name
    dists = resolver(DistributionFinder.Context(name=name))
  File "/home/smg/.pyenv/versions/quill_3.8.4/lib/python3.8/site-packages/importlib_metadata/__init__.py", line 417, in find_distributions
    found = self._search_paths(context.pattern, context.path)
AttributeError: 'Context' object has no attribute 'pattern'

**Comment by marcelotduarte** [comment]
You can continue in this issue. Later I'll take a look. It is related to a patch in the develop version. Or maybe to some bug in py 3.8.4 (I tested with 3.7 and 3.8.2). New python 3.8.5 release 2 days ago because 3.8.4 has bugs.

**Comment by elguavas** [comment]
interesting. in the meantime i'll create a new pyenv build with 3.8.5 and see if the problem still occurs. that might help clear up if it's a python bug. ;)

**Comment by elguavas** [comment]
ok, just to confirm, i just built a python pyenv python 3.8.5 and installed the devel cx-freeze in it, on the more complex project freezing give the same error as above. so it seems it wasn't a bug in python 3.8.4 causing the above crash.

**Comment by marcelotduarte** [comment]
What is your version of importlib_metadata?
Can you try to uninstall it and test? In my test in py38 I used the standard library importlib.metadata

**Comment by elguavas** [comment]
how to i get the version of that?

**Comment by marcelotduarte** [comment]
pip list|grep importlib

**Comment by elguavas** [comment]
just to be clear, i assume i'm using the standard importlib_metadata from the 3.8.5 build i made.

anyway pip reports 0.23

**Comment by marcelotduarte** [comment]
upgrade it or uninstall
the lastet version is 1.7.0

**Comment by marcelotduarte** [comment]
From the docs:
When imported on Python 3.8 and later, importlib_metadata replaces the DistributionFinder behavior from the stdlib, but leaves the API in tact.

**Comment by elguavas** [comment]
that's wierd... i intalled all the requirements.txt in the new virtualenv with -U.

**Comment by marcelotduarte** [comment]
From your traceback, it call the lib:
File "/home/smg/.pyenv/versions/3.8.4/lib/python3.8/importlib/metadata.py", line 172, in from_name
dists = resolver(DistributionFinder.Context(name=name))
File "/home/smg/.pyenv/versions/quill_3.8.4/lib/python3.8/site-packages/importlib_metadata/init.py", line 417, in find_distributions
found = self._search_paths(context.pattern, context.path)
AttributeError: 'Context' object has no attribute 'pattern'

**Comment by elguavas** [comment]
hmm, maybe something eles i installed from pip called in that weird version as a dependency...

**Comment by elguavas** [comment]
ok, i simply uninstalled the pip installed importlib-metadata and all now works as expected. thanks.

now i just have to work out why i had that pip installed version of 0.23 there in the first place.

**Comment by marcelotduarte** [comment]
You can force the installation of the most recent version. Maybe I force it as depedency in cx-freeze. Currently it is dependency only in py < 3.8

**Comment by elguavas** [comment]
ah ok, so cx-freeze pulled that in as a dependency?

**Comment by marcelotduarte** [comment]
No, for py38, no.

**Comment by elguavas** [comment]
i just wonder why pip install -U -r reqirements.txt didn't force that to upgrade to the newest version...

**Comment by elguavas** [comment]
yes it was probably in there since python 3.7 this project has been in existence a long time and would have used cx-freeze since years ago.

**Comment by marcelotduarte** [comment]
Run pip install --help and see the option --upgrade-strategy

**Comment by elguavas** [comment]
a new problem with the development version of cx-freeze now. on non-developement version this project build fine but with development version it builds but at runtime i'm getting:

qt.qpa.plugin: Could not find the Qt platform plugin "xcb" in ""
This application failed to start because no Qt platform plugin could be initialized. Reinstalling the application may fix this problem.

this project uses pyside2. i remember there used to be a problem with this a while back and involved manually adding the qt platform plugins in the build script. but that wasn't necessary for some time with the regular pip cx-freeze. it seems to be a problem again with the development build.

**Comment by marcelotduarte** [comment]
Yes, the newer versions of pyside changed something. There are issues about this. #665 is one.

**Comment by elguavas** [comment]
yeah --upgrade-strategy wouldn't have helped here i think because that importlib-metadata was left over from an old cx-freeze install and the new version on 3.8.x doesn't have it as a dependency right? so i'm not sure what pip's dependency resolution mechanism does in that case...

**Comment by elguavas** [comment]
> Yes, the newer versions of pyside changed something. There are issues about this. #665 is one.

ok, it's weird though that there is no runtime error when buil with current pip cx-freeze, only when built with development version.

**Comment by marcelotduarte** [comment]
no, cx-freeze do not have this as dependency. In the development version that I have:
install_requires = ["importlib_metadata; python_version < '3.8'"],

**Comment by elguavas** [comment]
right, but an older version of cx-freeze did have it as a dependency, right?

**Comment by marcelotduarte** [comment]
cx-freeze 6.2 has no dependency.

**Comment by elguavas** [comment]
yeah, but what i'm saying is, this project has used cx-freeze for years, since before 6.2, so an earlier version of cx-freeze could have pulled that in via pip as a dependency.

**Comment by marcelotduarte** [comment]
cx-freeze 6.1 has no dependency, but can use setuptools if installed.
cx-freeze 6.2 use setuptools always.
cx-freeze dev use setuptools and importlib_metadata (only for py < 3.8)

**Comment by elguavas** [comment]
so when cx-freeze via pip was upgraded from pre-6.2 to 6.2 that old version of  importlib-metadata would not have been removed.

**Comment by marcelotduarte** [comment]
If you run pip install -U -r requirements.txt you will see the resolution of dependencies.

**Comment by elguavas** [comment]
well anyway, i think it was pulled in as a dependency by _some_ old pip install, there has been no reason to install importlib-metadata via pip by itself for this project. so it could have been sitting there for ages in requirements.txt just getting re-installed on every python version update for the project.

if it was a dangling dependency though i just don't understand why pip --install -U -r requirement.txt didn't upgrade it to a newer version.

anyway i guess this is getting off topic.

**Comment by elguavas** [comment]
so the only issue i still have here is why there is no qt runtime error for this project when built with current pip cx-freeze, only when built with development version of cx-freeze.

**Comment by marcelotduarte** [comment]
with pip install cx-freeze==6.2 no runtime error?
and with pip install -U git+https://github.com/marcelotduarte/cx_Freeze.git@develop - runtime error?

**Comment by elguavas** [comment]
well i just tested with pip install cx-freeze==6.2 and there was no runtime error with the build.

now i'll test with development cx-freeze.

**Comment by marcelotduarte** [comment]
ah, ok, no build error.

**Comment by elguavas** [comment]
ok now it tested with pip install -U git+https://github.com/marcelotduarte/cx_Freeze.git@develop
and it builds but i get the qt runtime error.

**Comment by marcelotduarte** [comment]
Have a another option: pip install -U git+https://github.com/marcelotduarte/cx_Freeze.git@master

**Comment by elguavas** [comment]
ok will test now.

**Comment by elguavas** [comment]
ok, we hit the jackpot there. with  pip install -U git+https://github.com/marcelotduarte/cx_Freeze.git@master
the build ran, and there was no runtime error on app launch.

what's more however, i looked in the app's about box, which shows among other things the python version, and it's now showing 3.8.5 !! so it isn't picking up the system python version but the version it is built with, as expected.

so did that cx-freeze build incorporate your lib version fixes too?

**Comment by marcelotduarte** [comment]
master has 4 patches after 6.2 and develop has more 3

**Comment by marcelotduarte** [comment]
I removed 1 patch from develop, that I think that is the problem, can you test it again?

**Comment by elguavas** [comment]
ok. i was wondering specifically about the patches that might affect that "pulling in the right libs version" issue that we started with here. whatever is in current master fixes that issue for my current use case at least. and what's more master does not have the regression with qt platform package handling.

**Comment by elguavas** [comment]
> I removed 1 patch from develop, that I think that is the problem, can you test it again?

ahh yep, will try it now.

**Comment by elguavas** [comment]
yep, develop now builds the project and it runs without runtime error. also the sys.version is correct.

**Comment by marcelotduarte** [comment]
good. now 2 patches can be put in the master ;-)

**Comment by elguavas** [comment]
excellent! so does that make it closer to a new general pip release for you?

**Comment by marcelotduarte** [comment]
the third is an optimization, I'm trying to reduce the number of files in packages like pyqt5...

**Comment by elguavas** [comment]
ah, that makes sense then. i'm sure the situation is very similar for pyside2. we switched this project from pyqt to pyside2 about six months ago now.

**Comment by marcelotduarte** [comment]
yes, soon...

**Comment by elguavas** [comment]
cool. in the meantime i'll just keep using master for our test builds.

**Comment by elguavas** [comment]
but on the pyqt pyside2 side of things, it seems we still choke at runtime for missing platform packages at least.

**Comment by elguavas** [comment]
do you want me to close this issue now or is there something else you need tested?

**Comment by elguavas** [comment]
> but on the pyqt pyside2 side of things, it seems we still choke at runtime for missing platform packages at least.

or rather, what qt calls "platform plugins" were the issue here.

**Comment by marcelotduarte** [comment]
There are issues about plugins. If you have another issue, please open.

**Comment by cainesi** [comment]
Hi @elguavas.  Do you still have a copy of the frozen script where it was reporting the wrong python version? 

> pyenv installed python 3.8.4 , cx-freeze 6.2
> 
> hi there, if i freeze even a basic script using a pyenv installed version of python, when the script runs it reports sys.version of the system python... why??

**Comment by elguavas** [comment]
> Hi @elguavas. Do you still have a copy of the frozen script where it was reporting the wrong python version?
> 
> > pyenv installed python 3.8.4 , cx-freeze 6.2
> > hi there, if i freeze even a basic script using a pyenv installed version of python, when the script runs it reports sys.version of the system python... why??

it's at the bottom of the original post. even that 4 line script would show the error when it was occurring.

**Comment by cainesi** [comment]
Hi @elguavas.  Thanks, I saw the script there.  What I was wondering is if you still had a copy of the frozen bundles that cx_freeze created that did not work (probably not, but I thought it was worth checking--if you also had a copy of the build directory, that would be even better).  I would be interested to see what cx_freeze had done to the internal linking information in the binaries.

**Comment by elguavas** [comment]
ah ok, sorry i only have the current build dirs that do work using cx-freeze master, the old one are deleted.

**Comment by cainesi** [comment]
I figured.

If you wanted to check-out an old version of cx_Freeze and create another broken app bundle, I'd be interested to see it.  But not urgent.

**Comment by Fox214** [comment]
I have --enable-shared in my config when building, but I am getting the wrong patch/micro (its +0-3 from the real number in all my cases)
+0 is correct, +1 or +2 seem most common, I had at least one +3.

I commented here to repeat with a simple script: https://stackoverflow.com/questions/52359805/is-sys-version-info-reliable-for-python-version-checking/78425277#78425277

I'm not sure what is causing it though, any clues? (note printing it from setup.py is correct, so its only after its frozen its different)

**Comment by marcelotduarte** [comment]
The Python version shown will be the same Python version used to build cx_Freeze, not the Python version you use cx_Freeze to compile your applications. For example, I use manylinux's Python to distribute the binaries, which is at v 3.10.13, and on my local machine it is 3.10.12 to compile my apps, so 3.10.13 will appear in the sys.version of the built app.

**Comment by Fox214** [comment]
Are you saying that if I use a whl from pypy it will report the version used to build the wheel? (not the version I have locally)

I would need to download the source code and build myself to get it to match?

**Comment by marcelotduarte** [comment]
Yes!
Yes!
